#693 and #660 and #698

rmandelb opened the pull request:
This branch includes the following updates:

- The ability to initialize an InterpolatedImage from a specific HDU of a fits file (#660).
- Some updates to the COSMOSCatalog and the config interface to allow pre-selection of galaxies being used for simulations, either using parametric fits or realistic galaxies (#693).
- Some updates to the catalogs themselves to enable that pre-selection to happen correctly and efficiently (which means that taking advantage of some updates requires re-downloading the catalogs; GalSim will identify whether that is needed, and the ```galsim_download_cosmos``` utility can be used for the downloading and unpacking).  Plus there's a new COSMOS catalog available, which isn't technically a GalSim update except that it is now the default one that GalSim will use unless the original one is specified, and there are some new kwargs that can be used to choose between the currently available catalogs.

Since both Mike and I worked on parts of this, it would be helpful if someone else like @jmeyers314 could take a look.  And I think @beckermr was interested in this, so perhaps he'd like to try it out?

rmjarvis commented:
I also did #698 on this branch: deprecating the config 'Ring' object type and putting this functionality in a stamp type (again called 'Ring') instead.

jmeyers314 commented:
>it would be helpful if someone else like @jmeyers314 could take a look

Sure thing.

jmeyers314 commented:
I just started looking this over by cleaning out my GalSim directory with:

```
scons -c
rm -rf .scon*
```

and then compiling with plain `scons`.

For quite some time now (certainly nothing specific to this PR), this has been generating warnings on my laptop; all of which look like:

```
1 warning generated.
In file included from pysrc/CorrelatedNoise.cpp:28:
In file included from include/galsim/CorrelatedNoise.h:29:
In file included from /Users/josh/miniconda2/opt/lsst/tmv/include/TMV_Sym.h:27:
In file included from /Users/josh/miniconda2/opt/lsst/tmv/include/TMV.h:30:
In file included from /Users/josh/miniconda2/opt/lsst/tmv/include/tmv/TMV_Matrix.h:495:
/Users/josh/miniconda2/opt/lsst/tmv/include/tmv/TMV_Permutation.h:465:57: warning: dependent nested
      name specifier 'VectorView<T, A>::' for friend class declaration is not supported; turning off
      access control for 'Permutation' [-Wunsupported-friend]
        friend inline VectorView<T,A>& VectorView<T,A>::sort(
                                       ~~~~~~~~~~~~~~~~~^
```

rmjarvis commented:
> For quite some time now (certainly nothing specific to this PR), this has been generating warnings on my laptop; all of which look like:

Yes.  Sorry.  I know about those.  clang added this warning recently, and I haven't gotten around to fixing TMV to avoid it yet.  Until I do, you'll need to compile with WARN=false unfortunately.

jmeyers314 commented on galsim/scene.py:
```diff
@@ -33,7 +33,8 @@
 class COSMOSCatalog(object):
     """
-    A class representing a sample of galaxies from the COSMOS sample with F814W<23.5.
+    A class representing a random subsample of galaxies from the COSMOS sample with F814W<25.2
```
Should probably indicate at which galsim version number the default catalog changes, i.e. the meaning of "previous" here.

rmandelb replied:
Indeed.  Done!

jmeyers314 commented on galsim/scene.py:
```diff
@@ -86,46 +92,58 @@ class COSMOSCatalog(object):
+                            [default: None, which results in the same default as `file_name=None`.]
+    @param image_dir        Keyword that is only used for real galaxies, not parametric ones, to
+                            specify the directory of the image files.
+                            If a string containing no `/`, it is the relative path from the location
```
Is there a reason not to follow the normal *nix relative-vs-absolute path rule here?  I.e., relative paths can't *start* with '/', but could contain '/' at later positions?  I see there's even an os.path.isabs() utility for this.

rmandelb replied:
Hmmm, interesting question.   @rmjarvis , what are your thoughts?

rmjarvis replied:
I've always found the way these dir kwargs are parsed a bit confusing, to be honest.  I'd actually rather have just a single dir for everything and let the user use symlinks if they want to put things in different directories than the standard layout.  Seems to me that would be less confusing than having 3 different dir kwargs here.  Then there wouldn't be any need to distinguish between absolute and relative paths.

rmandelb replied:
I'm fine with that in principle.  But it might be tricky to do this in a way that doesn't break people's old code when they go to v1.4.  If you have ideas for how to cope with that, I would be fine with updating both COSMOSCatalog and RealGalaxyCatalog to this new model.

rmjarvis replied:
It would mean deprecating the old functionality, rather than removing it.  For now, probably just leave the code the same, but give deprecation warnings if either `image_dir` or `noise_dir` is not None.  I don't think there is any change required for the case where `dir` is given, but not `image_dir` or `noise_dir`.

rmandelb replied:
OK.  I will take a stab at it this weekend; will post when it�s ready to go over.

jmeyers314 commented on galsim/scene.py:
```diff
@@ -86,46 +92,58 @@ class COSMOSCatalog(object):
+                            the bulk of the I/O time is in the constructor.  If `preload=False`,
+                            there is approximately the same total I/O time (assuming you eventually
+                            use most of the image files referenced in the catalog), but it is spread
+                            over the various calls to getGal() and getPSF().  [default: False]
```
Should `getGal()` and `getPSF()` just be `makeGalaxy()` here?

rmandelb replied:
Yes they should.  That was clearly copied from the RealGalaxyCatalog description.  Fixed.

jmeyers314 commented on galsim/scene.py:
```diff
@@ -86,46 +92,58 @@ class COSMOSCatalog(object):
+                            there is approximately the same total I/O time (assuming you eventually
+                            use most of the image files referenced in the catalog), but it is spread
+                            over the various calls to getGal() and getPSF().  [default: False]
+    @param noise_dir        Keyword that is only used for real galaxies, not parametric ones.
```
Does `noise_dir` need any relative-vs-absolute path guidance like `image_dir`?

jmeyers314 commented on galsim/scene.py:
```diff
@@ -86,46 +92,58 @@ class COSMOSCatalog(object):
+    @param noise_dir        Keyword that is only used for real galaxies, not parametric ones.
+                            The directory of the noise files if different from the directory of the 
+                            image files.  [default: image_dir]
+    @param use_real         Enable the use of realistic galaxies?  [default: True]
```
Is there a speed or RAM benefit to having `use_real=False` when making parametric galaxies?  If so, then should add this to the documentation.  If not, then why have this kwarg?

rmandelb replied:
Well - in the very original version of the code I think there may have been more benefit because it was structured differently.  Now, there is a quite small benefit in terms of RAM and speed (don't have to read in the real catalog) and a possible logistical benefit, like if you never want to use realistic galaxies and deleted the real catalog from your system, you could still use COSMOSCatalog with ```use_real=False``` and not have it complain that it couldn't find the catalog.  I just added a note on this.

jmeyers314 commented on galsim/scene.py:
```diff
@@ -86,46 +92,58 @@ class COSMOSCatalog(object):
+                            not be allowed.
+    @param exclusion_level  Level of cuts to make on the galaxies based on the quality of postage
+                            stamp definition and/or parametric fit quality.  Options:
+                            "none": No cuts.
```
I assume there were some cuts in place when making the catalogs.  So should this be "No additional cuts", maybe?

rmandelb replied:
Well - those cuts were really minimal because we were trying to get as close as we could to a flux-limited sample.  But I suppose technically we should mention them and point to their description.  Done.

jmeyers314 commented on galsim/scene.py:
```diff
@@ -86,46 +92,58 @@ class COSMOSCatalog(object):
+    @param exclusion_level  Level of cuts to make on the galaxies based on the quality of postage
+                            stamp definition and/or parametric fit quality.  Options:
+                            "none": No cuts.
+                            "bad_ps": Apply cuts to eliminate galaxies that have failures in postage
```
I see `bad_ps` and immediately think that something is wrong with a power spectrum somewhere.  Suggest `bad_stamp` as an alternative.

rmandelb replied:
Fair enough.  Done.

jmeyers314 commented on galsim/scene.py:
```diff
@@ -194,47 +271,104 @@ def __init__(self, file_name=None, image_dir=None, dir=None, preload=False, nois
+        # If requested, select galaxies in a way that excludes suspect postage stamps (e.g., with
+        # deblending issues), suspect parametric model fits, or both of the above plus marginal
+        # ones.
+        if exclusion_level == 'bad_ps' or exclusion_level == 'marginal':
```
Looks like this is the same if statement as line 239.  Could these be streamlined?

rmandelb replied:
Good catch.  Done: https://github.com/GalSim-developers/GalSim/commit/7e3b9aa4889cd4d976daf7af47150bd4662abf3f

jmeyers314 commented on galsim/scene.py:
```diff
@@ -194,47 +271,104 @@ def __init__(self, file_name=None, image_dir=None, dir=None, preload=False, nois
+                           (self.selection_cat['average_mask_adjacent_pixel_count'] / \
+                               div_val < cut_ratio)) )
+
+        if exclusion_level == 'bad_fits' or exclusion_level == 'marginal':
```
Personally, I like the earlier style: `if exclusion_level in ['bad_fits', 'marginal']`.

rmandelb replied:
Fixed.

jmeyers314 commented on galsim/scene.py:
```diff
@@ -194,47 +271,104 @@ def __init__(self, file_name=None, image_dir=None, dir=None, preload=False, nois
             n = self.param_cat['sersicfit'][:,2]
-            mask &= ( (n < 5) | (hlr < 1./cosmos_pix_scale) ) 
-            # May add more cuts here if we discover other kinds of problematic objects.
+            mask &= ( (n < 5) | (hlr < 1./cosmos_pix_scale) )
```
I think this is saying that the HLR should be smaller than 1 arcsec.  This would be clearer with something like
```
hlr_pix = self.param_cat['sersicfit'][:, 1]
max_hlr_arcsec = 1.0
mask &= ((n < 5) | (hlr_pix < max_hlr_arcsec/cosmos_pix_scale))
```

jmeyers314 replied:
Also, I see down below that there's also a factor of `np.sqrt(q)` in the formula for hlr in arcsec.  Maybe a comment would be good here explaining why that's important down below but not here.

rmandelb replied:
Both the sqrt(q) issue and the nature of this cut itself are clarified in https://github.com/GalSim-developers/GalSim/commit/e7761cdd750d921e786282aae1349ee6942811f5

jmeyers314 commented on galsim/scene.py:
```diff
@@ -194,47 +271,104 @@ def __init__(self, file_name=None, image_dir=None, dir=None, preload=False, nois
+                    # The prefactor for n=4 is 3.607.  For n=1, it is 1.901.
+                    # It's not linear in these values, but for the sake of efficiency and the 
+                    # ability to work on the whole array at once, just linearly interpolate.
+                    # Hopefully, this can be improved as part of issue #693.  Maybe by storing the
```
I assume this comment is outdated?

rmandelb replied:
Yes, fixed.

jmeyers314 commented on galsim/scene.py:
```diff
@@ -401,7 +549,7 @@ def _makeParametric(self, indices, chromatic, sersic_prec, gsparams):
         gal_list = []
         for index in indices:
-            record = self.param_cat[self.orig_index[index]]
+            record = self.getParametricRecord(index)
```
Looks like I've been commenting on a fair amount of old code that just happened to show up in the imperfectly processed github diff.  Hope that's not too annoying.  I actually have a comment for some old code on line 751 in `getParametricRecord`, which isn't imperfectly diffed, so github won't let me comment on that line, so I'll place it here instead.  I was happy to see a generator expression:

```
record_dict = dict( ((k,record[k]) for k in record.dtype.names) )
```

I just wanted to point out that a generator expression passed as a single argument to a function doesn't need its own parens.  I.e., the following works and, at least to me, looks a bit cleaner:

```
record_dict = dict((k,record[k]) for k in record.dtype.names)
```

rmandelb replied:
Done.

jmeyers314 commented on galsim/scene.py:
```diff
@@ -350,13 +504,24 @@ def makeGalaxy(self, index=None, gal_type=None, chromatic=False, noise_pad_size=
-        # If deep, rescale the size and flux
+        # If trying to use the 23.5 sample and "fake" a deep sample, rescale the size and flux as
+        # suggested in the GREAT3 handbook.
         if deep:
```
Should the `deep` kwarg be disabled if using the 25.2 catalog?  (Same question for `_makeSingleGalaxy` down below).

rmandelb replied:
Yes.  Done.

jmeyers314 commented on examples/great3/cgc.yaml:
```diff
@@ -16,9 +16,11 @@
+# Note: This configuration will not produce images _identical_ to the ones used for Great3,
 # since the order of random number generation is different.  However, it is intended to
 # produce an _equivalent_ simulation as the one produced for Great3.
+modules:
```
I know that this example isn't really part of the tutorial series, but I think it would be nice for some more documentation on the `modules` top-level config type here.  (I think the only other example of this is `examples/des/draw_psf.yaml`.)  Or alternatively, maybe a new tutorial that more simply shows off the `modules` config type (or augment an existing tutorial?).

rmjarvis replied:
The DES examples are a more complete set of examples that use the modules field in various ways.  cf. the PR for #654.  I've called both the DES and the Great3 examples "Advanced Simulations" [here](https://github.com/GalSim-developers/GalSim/wiki/Tutorials) and mention the top-level modules field as a significant feature of both.

I'm not averse to including the modules in the main set of numbered demos.  I think the natural place will likely be demo13.  But that will have to wait until the chromatic stuff is folding into the config framework, which I haven't done yet.

jmeyers314 commented on examples/great3/cgc_faint.yaml:
```diff
@@ -0,0 +1,52 @@
+# using an I<25.2 sample (instead of faking a deeper sample using I<23.5 images).
+
+# Start with the cgc.yaml file, which we'll modify below.
+template: cgc.yaml
```
This top-level config type is also really cool.  We should advertise it in the tutorial series!

jmeyers314 commented on examples/great3/cgc.yaml:
```diff
@@ -122,16 +132,32 @@ gal:
+    # was based on running experiments with different seeing values, and checking how the S/N varied
+    # at fixed noise variance.  The results of those experiments were used to create a LookupTable
+    # in the GREAT3 scripts, but the formula below is pretty close.
+    fmax_var: '$0.004 * math.exp(2.9 * (1.1-psf_fwhm) * (0.95-psf_fwhm))'
```
Are the '$' and '@' Eval shorthands documented somewhere?

rmjarvis replied:
It's not live yet.  But yes.  Here: https://github.com/GalSim-developers/GalSim/wiki/Config-Values#shorthand-notation

jmeyers314 replied:
Great.  Thanks.

jmeyers314 commented:
Hi Rachel, Mike,

I've looked through most of the code now.  Looks good!  One caveat, I'm not as familiar with the config code, so you should only consider that "lightly" reviewed.

One last question/comment though:

I tried the following experiment to check on the tasks/jobs machinery.  From directory `$GALSIM_DIR/examples/great3/`, I ran (in three different terminals):

```
galsim cgc.yaml output.nfiles=10 output.dir=all10
galsim cgc.yaml output.nfiles=10 output.dir=5by5 -n 2 -j 1
galsim cgc.yaml output.nfiles=10 output.dir=5by5 -n 2 -j 2
```

My expectation was that the results in all10/ would bit-for-bit match those in 5by5/.  What I found was that all10/image-00[0-7]-0.fits matched 5by5/image-00[0-7]-0.fits, and similarly for the epoch_catalog files, but that numbers 8 and 9 in the series differed (rather significantly when I ds9'd the images).  It's as if a random number somewhere got skipped.  Is this surprising behavior?

rmjarvis commented:
> My expectation was that the results in all10/ would bit-for-bit match those in 5by5/. What I found was that all10/image-00[0-7]-0.fits matched 5by5/image-00[0-7]-0.fits, and similarly for the epoch_catalog files, but that numbers 8 and 9 in the series differed (rather significantly when I ds9'd the images). It's as if a random number somewhere got skipped. Is this surprising behavior?

Yes, that's probably indicating a bug.  Thanks for checking that.  I'll investigate and see if I can figure out what is going on there.  They should have been identical.

rmjarvis commented:
> numbers 8 and 9 in the series differed (rather significantly when I ds9'd the images). It's as if a random number somewhere got skipped. Is this surprising behavior?

I figured it out.  The problem was with the cosmos_cat.  The number of objects being rejected was coming out wrong.  I tracked it down to where `psf_fwhm` gets the current value of `psf.items.0.fwhm`.  Querying this when it hadn't already been calculated meant that it had to calculate it then, which should have been fine, but it jumped over the place where we tell it to use `index_key : image_num` (at the top of `psf`).

TL;DR, I fixed it here: 44ca3067cb70ebaa99375b09cafb01dfc2624dc8

rmandelb commented:
I wonder if there is a short/quick unit test we can add that will be able to catch the kind of issue that Josh found?

rmandelb commented:
I deprecated ```image_dir``` and ```noise_dir```, and updated the documentation to reflect this, in https://github.com/GalSim-developers/GalSim/commit/56e04d0c6b3b84a0cf493198d36dfd04d5d448c8 . This caused an issue for ```__repr__``` and therefore for pickling, which I fixed in a subsequent commit (I hope...).

@rmjarvis , @jmeyers314 , can you please check this out and see what you think?

rmjarvis commented:
You don't need to split off the directory in the `__repr__` function.  Providing just a catalog name, complete with the directory, will do that correctly already.  cf. 7c30fdf

rmandelb commented:
Good point.  Thanks.

rmjarvis commented:
> I wonder if there is a short/quick unit test we can add that will be able to catch the kind of issue that Josh found?

Good idea.  Done.  	bb87871

rmjarvis commented:
I think maybe all the comments have been dealt with, except for the suggestion to add template and modules examples to the demos.  Like I said, I think those will be most appropriate in demo12 and/or 13, so I'd like to push that into issue #510.

If that's ok with you (@jmeyers314 and @rmandelb), then I think we're ready to merge this?

jmeyers314 commented:
>I think we're ready to merge this?

Yep.  Looks good to me.  All tests pass.

rmandelb commented:
+1
